CamDetection.as

```diff
@@ -0,0 +1,146 @@
+void RenderMenuMain()
+{
+	if (!Setting_DebugShowActiveCamInMenu) {
+		return;
+	}
+	UI::BeginDisabled();
+	UI::MenuItem("Active Cam: " + tostring(g_activeCam));
+	UI::EndDisabled();
+}
+
+void UpdateActiveGameCam()
+{
+	g_activeCam = ActiveCam::None;
+
+	const Reflection::MwClassInfo@ tmTy = Reflection::GetType("CTrackMania");
+	const Reflection::MwMemberInfo@ gameSceneMember = tmTy.GetMember("GameScene");
+	uint GameCameraNodOffset = gameSceneMember.Offset + 0x10;
+
+	auto gameCameraNod = Dev::GetOffsetNod(GetApp(), GameCameraNodOffset);
+
+	// This is null in the menu and when cameras aren't being used.
+	if (gameCameraNod is null) {
+		return;
+	}
+
+	auto camModelNod = Dev::GetOffsetNod(gameCameraNod, 0x58);
+	auto camControlNod = Dev::GetOffsetNod(gameCameraNod, 0x68);
+
+	// Always 4 when backwards, and seemingly always 0 otherwise
+	bool isBackwards = Dev::GetOffsetUint32(gameCameraNod, 0xB0) == 0x4;
+
+	if (isBackwards) {
+		g_activeCam = ActiveCam::Backwards;
+	} else if (camModelNod !is null) {
+		auto ty = Reflection::TypeOf(camModelNod);
+		switch (ty.ID) {
+			// CPlugVehicleCameraRace2Model
+			case 0x90f6000: {
+				g_activeCam = TestCam1Alt(camModelNod) ? ActiveCam::Cam1Alt : ActiveCam::Cam1;
+				break;
+			}
+			// CPlugVehicleCameraRace3Model
+			case 0x90ef000: {
+				g_activeCam = TestCam2Alt(camModelNod) ? ActiveCam::Cam2Alt : ActiveCam::Cam2;
+				break;
+			}
+			// CPlugVehicleCameraInternalModel
+			case 0x90f7000: {
+				g_activeCam = TestCam3Alt(camModelNod) ? ActiveCam::Cam3Alt : ActiveCam::Cam3;
+				break;
+			}
+			default: {
+				trace('1 Got cam of unknown type: ' + ty.ID + ', ' + ty.Name);
+				UI::ShowNotification('1 Got cam of unknown type: ' + ty.ID + ', ' + ty.Name);
+				g_activeCam = ActiveCam::Other;
+			};
+		}
+	} else if (camControlNod !is null) {
+		auto ty = Reflection::TypeOf(camControlNod);
+		switch (ty.ID) {
+			// CGameControlCameraFree
+			case 0x306d000: {
+				g_activeCam = ActiveCam::FreeCam;
+				break;
+			}
+			// CGameControlCameraEditorOrbital
+			case 0x3125000: {
+				g_activeCam = ActiveCam::EditorOrbital;
+				break;
+			}
+			// CGameControlCameraOrbital3d
+			case 0x306e000: {
+				g_activeCam = ActiveCam::Orbital3d;
+				break;
+			}
+			// CGameControlCameraHelico
+			case 0x32f0000: {
+				g_activeCam = ActiveCam::Helico;
+				break;
+			}
+			// CGameControlCameraHmdExternal
+			case 0x3277000: {
```

Comment:
On Turbo, this ID is `0x0325F000`! Note that type ID's **can change over time**. We've seen this happen a few times. Some are more likely to change than others (for example if they are serialized they are unlikely to change). As such, comparing the type ID like this might not be realistic.

Reply:
Ahh cool. I used the ID instead of the class name for performance, but probably not worth the risk w/ IDs changing.

Reply:
We do have hashes of class & member names available for these kinds of cases that I use around Openplanet a bit, so we might be able to use that, but I'll need to think about it a bit before exposing those to the API.

Reply:
In the latest version I get all the class IDs from `Reflection::` on load and compare to those.